documents/Design.md

```diff
@@ -0,0 +1,25 @@
+# Dataset Viewer Design
+
+Programming Language: Python 3  
+Coding Style: PEP8  
+Design Pattern: Model View Presenter (Passive View)
+## Class Diagram
+![Class Diagram](ClassDiagram.png)  
+DimensionView: Contains the slider, buttons, and stepper for an individual dimension.  
+DimensionPresenter: Manages the behavior of the DimensionView.  
+MainView: Contains the individual DimensionViews and the central plot.  
+MainPresenter: Manages the behavior of the MainView and retrieves data from the MainModel.  
+MainModel: Contains the data array. 
+## Sequence Diagrams
+### Pressing an X Button
+![X Button Press Sequence Diagram](XButtonPress.png)
+### Stepper Change
+![Stepper Change Sequence Diagram](StepperChange.png)
+## Development Principles
+* Testing must take place alongside development
+* Reviewing must take place before code is merged into the master branch
+* The reviewer must be someone other than the developer
+* Merging can only take place if all tests pass and the reviewer(s) finds the code satisfactory
+## Testing and QA
+* Use the `unittest` library
+* `mock` the View
```

Comment:
You are going to measure test coverage. I would say that having 85% of the code base covered by tests should be a minimum threshold. MSlice achieves ~90%

Reply:
All tests should be executed by the CI build server

Reply:
All tests should be written to the same quality as the rest of the codebase.

Reply:
Nothing can be merged without builds passing.

Reply:
Tests must not be changed without good justification. i.e. tests cannot be made to pass if they represent legitimate failures.

Reply:
There should be destructive as well as happy-path tests put in place.